Percentile Function

dhruvsaxena11 opened the pull request:
Hi,

Close #17469

I have added the percentile function in statistical functions of numpy frontend.

Please kindly check and review it.

Thank You

Dhruv Saxena

ivy-leaves commented:
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

dhruvsaxena11 commented:
@ivy-leaves 

I have updated the description please kindly check.

Thank You

Dhruv Saxena

dhruvsaxena11 commented:
@ivy-leaves 

please kindly check my PR.

Thank You

dhruvsaxena11 reviewed (commented):
@ivy-leaves 
Plz kindly review my PR as I updated the description. 

Thank you

zhumakhan commented:
Hi, thanks for you PR!
Please also add test and decorators to the function like @to_ivy_arrays_and_back

dhruvsaxena11 commented:
@zhumakhan 

Hi,

I have updated the files and added the decorators to function and added the test.
please kindly check and suggest.

Thank You

dhruvsaxena11 reviewed (commented):
@zhumakhan 

please kindly check and merge

zhumakhan commented:
Could you please fix the implementation? Decorators in percentile.py file are not imported
Thanks!

dhruvsaxena11 commented:
@zhumakhan I updated the code please kindly check.

Thank You

dhruvsaxena11 commented:
@zhumakhan please kindly check and merge

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -35,6 +43,41 @@ def _cpercentile(N, percent, key=lambda x: x):
+    elif axis == 0:
+        nanlessarrayofarrays = ivy.where(ivy.isnan(a), ivy.inf, a)
+        nanlessarrayofarrays = ivy.swapaxes(nanlessarrayofarrays, 0, 1)
+        resultarray = ivy.percentile(nanlessarrayofarrays, q, axis=0, keepdims=keepdims)
```
It seems `ivy.percentile` does not exist

dhruvsaxena11 replied:
I have made some changes

zhumakhan commented on ivy/functional/frontends/numpy/statistics/percentile.py:
```diff
@@ -0,0 +1,44 @@
+@to_ivy_arrays_and_back
+@from_zero_dim_arrays_to_scalar
+
+def percentile(
```
Defining `percentile`  in order_statistics.py file is enough. No need to duplicate.

dhruvsaxena11 replied:
Done

ivy-seed commented:
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

dhruvsaxena11 commented:
Will working on it

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -1,5 +1,16 @@
+)
+
 from ivy.functional.frontends.numpy.func_wrapper import (
     to_ivy_arrays_and_back,
```
Could you please remove this duplicate import?

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -37,6 +48,82 @@ def _cpercentile(N, percent, key=lambda x: x):
     d0 = key(N[int(f)]) * (c - k)
     d1 = key(N[int(c)]) * (k - f)
     return d0 + d1
+import ivy
```
duplicated import of ivy

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -1,5 +1,16 @@
 # global
 import ivy
+import numpy
```
please remove unused imports

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -1,5 +1,16 @@
 # global
 import ivy
+import numpy
+import ivy.numpy as np 
```
This does not seem to exist

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -37,6 +48,82 @@ def _cpercentile(N, percent, key=lambda x: x):
+
+
+@ivy.to_ivy_arrays_and_back
+@ivy.handle_numpy_out
```
`@handle_numpy_out` should work. `ivy` does not have `handle_numpy_out` attribute

dhruvsaxena11 commented:
@zhumakhan 

I have updated the code. Please kindly check and merge.

zhumakhan commented on ivy/functional/frontends/numpy/statistics/order_statistics.py:
```diff
@@ -39,6 +43,82 @@ def _cpercentile(N, percent, key=lambda x: x):
+def _quantile_is_valid(q):
+    return ivy.all(ivy.logical_and(q >= 0, q <= 1))
+
+def _cpercentile(arr, q, interpolation='linear'):
```
It seems `_cpercentile` function already exists: https://github.com/unifyai/ivy/blob/main/ivy/functional/frontends/numpy/statistics/order_statistics.py#L13

zhumakhan commented:
Hi, many tests are failing due to various reasons like shape related, unsupported dtypes and missing arguments to test function. Could you please thoroughly go through tests and fix them?
Thanks!

ivy-seed commented:
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
